File: pset2-c/readability/readability.c
// CS50 pset2: Readability (Coleman-Liau)
#include <cs50.h>
#include <stdio.h>
#include <ctype.h>

int main(void)
{
    string text = get_string("Text: ");

    int letters = 0, words = 0, sentences = 0;
    for (char *p = text; *p; ++p)
    {
        unsigned char c = (unsigned char)*p;
        if (isalpha(c))
            letters++;
        if (c == ' ')
            words++;
        if (c == '.' || c == '!' || c == '?')
            sentences++;
    }
    // Count the last word if input isn't empty
    if (text[0] != '\0')
        words++;

    // Avoid division by zero
    if (words == 0)
    {
        printf("Before Grade 1\n");
        return 0;
    }

    double L = (letters * 100.0) / words;
    double S = (sentences * 100.0) / words;
    double idx = 0.0588 * L - 0.296 * S - 15.8;

    int grade = (int)(idx >= 0 ? idx + 0.5 : idx - 0.5); // round to nearest

    if (grade < 1)
        printf("Before Grade 1\n");
    else if (grade >= 16)
        printf("Grade 16+\n");
    else
        printf("Grade %d\n", grade);

    return 0;
}

File: pset6-py/readability/readability.py
import string

def main():
    text = input("Text: ")
    
    # Count letters, words, and sentences
    letters = count_letters(text)
    words = count_words(text)
    sentences = count_sentences(text)
    
    # Calculate Coleman-Liau index
    L = letters / words * 100
    S = sentences / words * 100
    
    index = 0.0588 * L - 0.296 * S - 15.8
    grade = round(index)
    
    if grade < 1:
        print("Before Grade 1")
    elif grade >= 16:
        print("Grade 16+")
    else:
        print(f"Grade {grade}")

def count_letters(text):
    return sum(1 for c in text if c.isalpha())

def count_words(text):
    return len(text.split())

def count_sentences(text):
    return sum(1 for c in text if c in '.!?')

if __name__ == "__main__":
    main()
